File: python/src/threads/adapters/models/anthropic/stream.py
"""Messages API stream events to core chunks. Every event is parsed at this boundary.

Text streams as deltas at once. Parts are held until the stop reason is known: at max_tokens
the block being generated was cut off, and a cut-off tool call or thinking block never exists.
Thinking blocks are stored whole, signature included, before their part names
them.
"""

from collections.abc import Sequence
from dataclasses import dataclass, field
from typing import Literal, assert_never

from pydantic import JsonValue, ValidationError

from threads.adapters.models.anthropic.caching import Ttl
from threads.adapters.models.anthropic.request import PROVIDER
from threads.adapters.models.anthropic.wire import (
    ARGS,
    CITATION,
    STOPS,
    BlockDelta,
    BlockStart,
    BlockStop,
    CitationsDelta,
    Counts,
    Event,
    JsonDelta,
    Location,
    MessageDelta,
    MessageStart,
    MessageStop,
    SignatureDelta,
    StreamError,
    TextDelta,
    ThinkingDelta,
    ToolStart,
    Web,
)
from threads.adapters.models.render import UnsupportedContentError, store_json
from threads.log import (
    CallId,
    CitationPart,
    HostedToolPart,
    OutputPart,
    ReasoningPart,
    TextPart,
    ToolUsePart,
    Usage,
)
from threads.loop.model import Delta, Done, ModelChunk, ModelContext, PartChunk, StopReason


@dataclass
class _Block:
    start: dict[str, JsonValue]
    text: list[str] = field(default_factory=list[str])
    citations: list[CitationPart] = field(default_factory=list[CitationPart])
    signature: str = ""


@dataclass
class Assembler:
    """Folds one attempt's events into chunks."""

    context: ModelContext
    model: str
    documents: Sequence[str]
    ttl: Ttl | None
    """The pinned prompt_cache TTL: a cache write under any other is unknown."""
    open: dict[int, _Block] = field(default_factory=dict[int, _Block])
    parts: list[list[OutputPart]] = field(default_factory=list[list[OutputPart]])
    """Each finished block's parts, in block order."""
    usage: Counts = field(default_factory=Counts)
    stop: StopReason | None = None

    async def feed(self, event: Event) -> Sequence[ModelChunk]:
        match event:
            case MessageStart(message=message):
                self.usage = _merge(self.usage, message.usage)
            case BlockStart(index=index, content_block=block):
                self.open[index] = _Block(block)
            case BlockDelta(index=index, delta=delta):
                return self._delta(index, delta)
            case BlockStop(index=index):
                self.parts.append(await self._finish(self.open.pop(index)))
            case MessageDelta(delta=delta, usage=usage):
                self.usage = _merge(self.usage, usage)
                self.stop = STOPS.get(delta.stop_reason or "", "other")
            case MessageStop():
                return self._done()
            case StreamError(error=error):
                raise ProviderStreamError(error.type, error.message)
            case _:
                assert_never(event)
        return ()

    def _delta(
        self,
        index: int,
        delta: TextDelta | JsonDelta | ThinkingDelta | SignatureDelta | CitationsDelta,
    ) -> Sequence[ModelChunk]:
        block = self.open[index]
        match delta:
            case TextDelta(text=text):
                block.text.append(text)
                # Blocks close in order, so the text is the next part; with another block open
                # its index isn't known, and it arrives whole at commit.
                if len(self.open) > 1:
                    return ()
                return (Delta(sum(len(p) for p in self.parts), text),)
            case JsonDelta(partial_json=text) | ThinkingDelta(thinking=text):
                block.text.append(text)
            case SignatureDelta(signature=signature):
                block.signature += signature
            case CitationsDelta(citation=citation):
                block.citations.append(self._citation(citation))
            case _:
                assert_never(delta)
        return ()

    def _citation(self, raw: dict[str, JsonValue]) -> CitationPart:
        location = CITATION.validate_python(raw)
        match location:
            case Web(url=url, title=title, cited_text=cited):
                return _cite("web", url, title, cited)
            case Location(document_index=index, document_title=title, cited_text=cited):
                return _cite("document", self.documents[index], title, cited)
            case _:
                assert_never(location)

    async def _finish(self, block: _Block) -> list[OutputPart]:
        kind = block.start.get("type")
        text = "".join(block.text)
        match kind:
            case "text":
                return [TextPart(type="text", text=text), *block.citations]
            case "tool_use":
                return _tool_use(block.start, text)
            case "thinking":
                whole: JsonValue = {
                    "type": "thinking",
                    "thinking": text,
                    "signature": block.signature,
                }
                return [await self._reasoning("thinking", whole)]
            case "redacted_thinking":
                return [await self._reasoning("redacted_thinking", block.start)]
            case str():
                return [await self._hosted(kind, block, text)]
            case _:
                raise UnsupportedContentError("content_unsupported", f"{kind} block")

    async def _hosted(self, kind: str, block: _Block, text: str) -> HostedToolPart:
        """A server tool use or result block, kept whole and never dispatched.
        A server_tool_use streams its input as JSON deltas; result blocks arrive whole."""
        whole = dict(block.start)
        if text:
            whole["input"] = ARGS.validate_json(text)
        name = whole.get("name")
        ref = await store_json(self.context, whole)
        return HostedToolPart(
            type="hosted_tool",
            provider=PROVIDER,
            model=self.model,
            format=kind,
            name=name if isinstance(name, str) else kind,
            ref=ref,
        )

    async def _reasoning(self, form: str, block: JsonValue) -> ReasoningPart:
        ref = await store_json(self.context, block)
        return ReasoningPart(
            type="reasoning", provider=PROVIDER, model=self.model, format=form, ref=ref
        )

    def _done(self) -> Sequence[ModelChunk]:
        stop = self.stop or "other"
        blocks = self.parts
        # At max_tokens the last block was still being generated: only text survives a cut.
        if stop == "max_tokens" and blocks and not _is_text(blocks[-1]):
            blocks = blocks[:-1]
        chunks: list[ModelChunk] = [PartChunk(p) for parts in blocks for p in parts]
        chunks.append(Done(stop, _usage(self.usage, self.ttl)))
        return chunks


def _is_text(parts: Sequence[OutputPart]) -> bool:
    return bool(parts) and isinstance(parts[0], TextPart)


class ProviderStreamError(Exception):
    """An `error` event inside the stream."""

    def __init__(self, kind: str, message: str) -> None:
        super().__init__(f"{kind}: {message}")
        self.kind = kind


def _tool_use(start: dict[str, JsonValue], text: str) -> list[OutputPart]:
    """A call whose input is not a whole JSON object was cut off: it never exists."""
    head = ToolStart.model_validate(start)
    try:
        args = ARGS.validate_json(text) if text else ARGS.validate_python(start.get("input"))
    except ValidationError:
        return []
    return [ToolUsePart(type="tool_use", call_id=CallId(head.id), name=head.name, input=args)]


def _cite(
    kind: Literal["web", "document"], source: str, title: str | None, cited: str
) -> CitationPart:
    if title is None:
        return CitationPart(type="citation", source_kind=kind, source_id=source, cited_text=cited)
    return CitationPart(
        type="citation", source_kind=kind, source_id=source, title=title, cited_text=cited
    )


def _merge(old: Counts, new: Counts) -> Counts:
    """message_delta repeats usage as running totals: the latest known value of each wins."""

    def pick(a: int | None, b: int | None) -> int | None:
        return a if b is None else b

    return Counts(
        input_tokens=pick(old.input_tokens, new.input_tokens),
        output_tokens=pick(old.output_tokens, new.output_tokens),
        cache_creation_input_tokens=pick(
            old.cache_creation_input_tokens, new.cache_creation_input_tokens
        ),
        cache_read_input_tokens=pick(old.cache_read_input_tokens, new.cache_read_input_tokens),
        cache_creation=old.cache_creation if new.cache_creation is None else new.cache_creation,
    )


def _usage(u: Counts, ttl: Ttl | None) -> Usage:
    """Anthropic's input_tokens already exclude cache reads and writes; its usage profile has
    both cache categories, so they are present, and None when they didn't arrive."""
    return Usage(
        input_tokens=u.input_tokens,
        output_tokens=u.output_tokens,
        cache_read_tokens=u.cache_read_input_tokens,
        cache_write_tokens=_priced_writes(u, ttl),
    )


def _priced_writes(u: Counts, ttl: Ttl | None) -> int | None:
    """Cache writes all billed at the pinned TTL's price, or None (unknown) when the provider
    reports any under the other TTL: a number priced at the wrong rate would understate cost."""
    split = u.cache_creation
    other = None
    if split is not None and ttl is not None:
        other = split.ephemeral_5m_input_tokens if ttl == "1h" else split.ephemeral_1h_input_tokens
    return None if (other or 0) > 0 else u.cache_creation_input_tokens

File: typescript/packages/anthropic/src/stream.ts
import type { Json, ModelChunk, Usage } from "@threads/core/adapter";
import { assertNever, JsonObject } from "@threads/core/adapter";
import { z } from "zod";
import { type BlockContext, blockParts } from "./blocks";
import type { Ttl } from "./caching";

// Messages API stream events → ModelChunks. Events are provider output, so each is parsed; an
// event type this adapter doesn't know is skipped, as the API asks clients to do.

const Tokens = z.int().nonnegative().nullish();
const ProviderUsage = z.object({
  input_tokens: Tokens,
  output_tokens: Tokens,
  cache_read_input_tokens: Tokens,
  cache_creation_input_tokens: Tokens,
  cache_creation: z
    .object({
      ephemeral_5m_input_tokens: Tokens,
      ephemeral_1h_input_tokens: Tokens,
    })
    .nullish(),
  output_tokens_details: z.object({ thinking_tokens: z.int() }).nullish(),
});
type ProviderUsage = z.infer<typeof ProviderUsage>;

const Delta = z.discriminatedUnion("type", [
  z.object({ type: z.literal("text_delta"), text: z.string() }),
  z.object({ type: z.literal("input_json_delta"), partial_json: z.string() }),
  z.object({ type: z.literal("citations_delta"), citation: JsonObject }),
  z.object({ type: z.literal("thinking_delta"), thinking: z.string() }),
  z.object({ type: z.literal("signature_delta"), signature: z.string() }),
]);

const Index = z.int().nonnegative();
const Event = z.discriminatedUnion("type", [
  z.object({
    type: z.literal("message_start"),
    message: z.object({ usage: ProviderUsage }),
  }),
  z.object({
    type: z.literal("content_block_start"),
    index: Index,
    content_block: JsonObject,
  }),
  z.object({
    type: z.literal("content_block_delta"),
    index: Index,
    delta: Delta,
  }),
  z.object({ type: z.literal("content_block_stop"), index: Index }),
  z.object({
    type: z.literal("message_delta"),
    delta: z.object({ stop_reason: z.string().nullish() }),
    usage: ProviderUsage,
  }),
  z.object({ type: z.literal("message_stop") }),
]);
const KNOWN = new Set<string>(Event.options.map((o) => o.shape.type.value));
const Tagged = z.object({ type: z.string() });

type Stop = Extract<ModelChunk, { kind: "done" }>["stop_reason"];
// A reason the SDK doesn't list yet is "other", which ends the turn with error, never success.
const STOPS = new Map<string, Stop>([
  ["end_turn", "end_turn"],
  ["tool_use", "tool_use"],
  ["max_tokens", "max_tokens"],
  ["stop_sequence", "stop_sequence"],
  ["refusal", "refusal"],
  ["pause_turn", "pause_turn"],
  ["model_context_window_exceeded", "context_window_exceeded"],
]);

type Open = { block: { [key: string]: Json }; json: string };

export async function* decode(
  events: AsyncIterable<unknown>,
  ctx: BlockContext,
  ttl: Ttl | undefined,
): AsyncGenerator<ModelChunk, void, undefined> {
  const open = new Map<number, Open>();
  let usage: ProviderUsage = {};
  let stop: Stop = "other";
  for await (const raw of events) {
    if (!KNOWN.has(Tagged.parse(raw).type)) continue;
    const e = Event.parse(raw);
    switch (e.type) {
      case "message_start":
        usage = e.message.usage;
        break;
      case "content_block_start":
        open.set(e.index, { block: { ...e.content_block }, json: "" });
        break;
      case "content_block_delta": {
        const text = apply(openAt(open, e.index), e.delta);
        if (text !== undefined) yield { kind: "delta", text };
        break;
      }
      case "content_block_stop":
        for (const part of await blockParts(close(openAt(open, e.index)), ctx))
          yield { kind: "part", part };
        open.delete(e.index);
        break;
      case "message_delta":
        usage = merge(usage, e.usage);
        stop = stopOf(e.delta.stop_reason);
        break;
      case "message_stop":
        yield { kind: "done", stop_reason: stop, usage: toUsage(usage, ttl) };
        return;
      default:
        assertNever(e);
    }
  }
  throw new Error("anthropic stream ended before message_stop");
}

function openAt(open: Map<number, Open>, index: number): Open {
  const block = open.get(index);
  if (block === undefined) throw new Error(`no open content block ${index}`);
  return block;
}

function stopOf(reason: string | null | undefined): Stop {
  return STOPS.get(reason ?? "") ?? "other";
}

/** Applies one delta to its open block; returns streamed text, if any. */
function apply(o: Open, delta: z.infer<typeof Delta>): string | undefined {
  const text = (key: string, more: string): void => {
    const was = o.block[key];
    o.block[key] = `${typeof was === "string" ? was : ""}${more}`;
  };
  switch (delta.type) {
    case "text_delta":
      text("text", delta.text);
      return delta.text;
    case "thinking_delta":
      text("thinking", delta.thinking);
      return undefined;
    case "signature_delta":
      o.block["signature"] = delta.signature;
      return undefined;
    case "citations_delta": {
      const was = o.block["citations"];
      o.block["citations"] = [
        ...(Array.isArray(was) ? was : []),
        delta.citation,
      ];
      return undefined;
    }
    case "input_json_delta":
      o.json += delta.partial_json;
      return undefined;
    default:
      return assertNever(delta);
  }
}

/** Streamed tool input arrives as JSON text; an empty stream leaves the start's input. */
function close(o: Open): { readonly [key: string]: Json } {
  return o.json === ""
    ? o.block
    : { ...o.block, input: JsonObject.parse(JSON.parse(o.json)) };
}

/** message_delta usage is cumulative; a field it leaves out keeps message_start's value. */
function merge(a: ProviderUsage, b: ProviderUsage): ProviderUsage {
  return {
    input_tokens: b.input_tokens ?? a.input_tokens,
    output_tokens: b.output_tokens ?? a.output_tokens,
    cache_read_input_tokens:
      b.cache_read_input_tokens ?? a.cache_read_input_tokens,
    cache_creation_input_tokens:
      b.cache_creation_input_tokens ?? a.cache_creation_input_tokens,
    cache_creation: b.cache_creation ?? a.cache_creation,
    output_tokens_details: b.output_tokens_details ?? a.output_tokens_details,
  };
}

/** Unknown is null, never 0. input_tokens already excludes cache. */
function toUsage(u: ProviderUsage, ttl: Ttl | undefined): Usage {
  return {
    input_tokens: u.input_tokens ?? null,
    output_tokens: u.output_tokens ?? null,
    cache_read_tokens: u.cache_read_input_tokens ?? null,
    cache_write_tokens: pricedWrites(u, ttl),
    reasoning_tokens: u.output_tokens_details?.thinking_tokens ?? null,
  };
}

/**
 * Cache writes all billed at the pinned TTL's price, or null (unknown) when the provider reports
 * any under the other TTL: a number priced at the wrong rate would understate the cost.
 */
function pricedWrites(u: ProviderUsage, ttl: Ttl | undefined): number | null {
  const other =
    ttl === "1h"
      ? u.cache_creation?.ephemeral_5m_input_tokens
      : ttl === "5m"
        ? u.cache_creation?.ephemeral_1h_input_tokens
        : undefined;
  return (other ?? 0) > 0 ? null : (u.cache_creation_input_tokens ?? null);
}
